pkg/yurthub/filter/approver.go

```diff
@@ -33,194 +33,247 @@ import (
+		return false
+	}
+
+	key := getKeyByRequest(req)
+	if len(key) == 0 {
+		return false
+	}
+
+	a.Lock()
+	defer a.Unlock()
+	if _, ok := a.reqKeyToName[key]; ok {
+		return true
 	}
+
+	return false
+}
+
+func (a *approver) GetFilterName(req *http.Request) string {
+	key := getKeyByRequest(req)
+	if len(key) == 0 {
+		return ""
+	}
+
 	a.Lock()
 	defer a.Unlock()
-	for _, requests := range a.nameToRequests {
-		for _, request := range requests {
-			if request.Equal(comp, resource, verb) {
+	return a.reqKeyToName[key]
+}
 
-				return true
+// Determine whether it is a whitelist resource
+func isWhitelistReq(req *http.Request) bool {
+	key := getKeyByRequest(req)
+	if _, ok := defaultWhiteListRequests[key]; ok {
+		return true
+	}
+	return false
+}
+
+func (a *approver) addConfigMap(obj interface{}) {
+	cm, ok := obj.(*corev1.ConfigMap)
+	if !ok {
+		return
+	}
+
+	// get reqKeyToName of user request setting from configmap
+	reqKeyToNameFromCM := make(map[string]string)
+	for key, setting := range cm.Data {
+		if name, ok := hasFilterName(key); ok {
+			for _, key := range parseRequestSetting(setting) {
+				reqKeyToNameFromCM[key] = name
 			}
 		}
 	}
-	return false
+
+	// update reqKeyToName by merging user setting
+	a.merge("add", reqKeyToNameFromCM)
 }
 
-func (a *approver) GetFilterName(req *http.Request) string {
-	ctx := req.Context()
-	comp, ok := util.ClientComponentFrom(ctx)
+func (a *approver) updateConfigMap(oldObj, newObj interface{}) {
+	oldCM, ok := oldObj.(*corev1.ConfigMap)
 	if !ok {
-		return ""
+		return
 	}
 
-	info, ok := apirequest.RequestInfoFrom(ctx)
+	newCM, ok := newObj.(*corev1.ConfigMap)
 	if !ok {
-		return ""
+		return
 	}
-	a.Lock()
-	defer a.Unlock()
-	for name, requests := range a.nameToRequests {
-		for _, request := range requests {
-			if request.Equal(comp, info.Resource, info.Verb) {
-				return name
+
+	// request settings are changed or not
+	needUpdated := requestSettingsUpdated(oldCM.Data, newCM.Data)
+	if !needUpdated {
+		return
+	}
+
+	// get reqKeyToName of user request setting from new configmap
+	reqKeyToNameFromCM := make(map[string]string)
+	for key, setting := range newCM.Data {
+		if name, ok := hasFilterName(key); ok {
+			for _, key := range parseRequestSetting(setting) {
+				reqKeyToNameFromCM[key] = name
 			}
 		}
 	}
-	return ""
+
+	// update reqKeyToName by merging user setting
+	a.merge("update", reqKeyToNameFromCM)
 }
 
-// Determine whether it is a whitelist resource
-func (a *approver) isWhitelistReq(comp, resource, verb string) bool {
-	for _, req := range a.whiteListRequests {
-		if req.Equal(comp, resource, verb) {
-			return true
+// merge is used to add specified setting into reqKeyToName map.
+func (a *approver) merge(action string, keyToNameSetting map[string]string) {
+	a.Lock()
+	defer a.Unlock()
+	// remove current user setting from reqKeyToName and left default setting only
+	for key := range a.reqKeyToName {
+		if _, ok := defaultReqKeyToName[key]; !ok {
```

Comment:
This code may be som confuse.  Delete not default keys mean only use default keys.
```go
	newReqKey := map[string]string{}
	for key, name := range defaultReqKeyToName {
		newReqKey[key] = name
	}
	for key, name := range keyToNameSetting {
		// if filter setting is duplicated, only recognize the first setting.
		if currentName, ok := newReqKey[key]; !ok {
			newReqKey[key] = name
		} else {
			klog.Warningf("request %s has already used filter %s, so skip filter %s setting", key, currentName, name)
		}
	}
	a.Lock()
	defer a.Unlock()
	a.reqKeyToName = newReqKey
```